runtime-sdk/modules/evm/src/precompile/confidential.rs

```diff
@@ -0,0 +1,289 @@
+    if input.len() < 128 {
+        return Err(PrecompileFailure::Error {
+            exit_status: ExitError::Other("input length must be at least 128 bytes".into()),
+        });
+    }
+
+    key.copy_from_slice(&input[..32]);
+    nonce.copy_from_slice(&input[32..(32 + NONCE_SIZE)]);
+
+    text_len.copy_from_slice(&input[64..96]);
+    let text_len_big = BigUint::from_bytes_be(&text_len);
+    let text_len = text_len_big
+        .to_usize()
+        .ok_or_else(|| PrecompileFailure::Error {
+            exit_status: ExitError::Other("text length out of bounds".into()),
+        })?;
+    let text_size = text_len.saturating_add(31) & (!0x1f); // Round up to 32 bytes.
+
+    ad_len.copy_from_slice(&input[96..128]);
+    let ad_len_big = BigUint::from_bytes_be(&ad_len);
+    let ad_len = ad_len_big
+        .to_usize()
+        .ok_or_else(|| PrecompileFailure::Error {
+            exit_status: ExitError::Other("additional data length out of bounds".into()),
+        })?;
+    let ad_size = ad_len.saturating_add(31) & (!0x1f); // Round up to 32 bytes.
+    if input.len() < 128 + ad_size + text_size {
+        return Err(PrecompileFailure::Error {
+            exit_status: ExitError::Other("input too short".into()),
+        });
+    }
+
+    let text = &input[128..(128 + text_len)];
+    let ad = &input[(128 + text_size)..(128 + text_size + ad_len)];
+
+    Ok((key, nonce, text, ad))
+}
+
+pub(super) fn call_deoxysii_seal(
+    input: &[u8],
+    target_gas: Option<u64>,
+    _context: &Context,
+    _is_static: bool,
+) -> PrecompileResult {
+    let gas_cost = linear_cost(target_gas, input.len() as u64, 10_000, 1000)?;
+
+    // Input encoding: bytes32 key || bytes32 nonce || uint plaintext_len || uint ad_len || plaintext || ad.
+    let (key, nonce, text, ad) = decode_deoxysii_call_args(input)?;
+
+    let deoxysii = DeoxysII::new(&key);
+    let encrypted = deoxysii.seal(&nonce, text, ad);
+
+    let result = PrecompileOutput {
+        exit_status: ExitSucceed::Returned,
+        cost: gas_cost,
+        output: encrypted,
+        logs: Default::default(),
+    };
+    Ok(result)
+}
+
+pub(super) fn call_deoxysii_open(
+    input: &[u8],
+    target_gas: Option<u64>,
+    _context: &Context,
+    _is_static: bool,
+) -> PrecompileResult {
+    let gas_cost = linear_cost(target_gas, input.len() as u64, 10_000, 1000)?;
+
+    // Input encoding: bytes32 key || bytes32 nonce || uint ciphertext_len || uint ad_len || ciphertext || ad.
+    let (key, nonce, ciphertext, ad) = decode_deoxysii_call_args(input)?;
+    let ciphertext = ciphertext.to_vec();
+
+    let deoxysii = DeoxysII::new(&key);
+    let result = match deoxysii.open(&nonce, ciphertext, ad) {
+        Ok(decrypted) => PrecompileOutput {
+            exit_status: ExitSucceed::Returned,
+            cost: gas_cost,
+            output: decrypted,
+            logs: Default::default(),
+        },
+        Err(_) => PrecompileOutput {
+            exit_status: ExitSucceed::Returned,
```

Comment:
This should cause a reversion. The data, in fact, could be empty, so that's not a reliable differentiator.